### PR TITLE
[text] refine tokenizer

### PR DESCRIPTION
**Pull request by Mddct**
issues: https://github.com/wenet-e2e/wenet/issues/2160

in this pr:
- [x] tokenzier interface
- [x] whisper tokenizer
- [x] wenet tokenizer
- [x] split WenetTokenizer (WenetTokenizer 作为测试对齐）
   - [x]  CharTokenizer
   - [x]  BpeTokenizer
- [x] check if can work with multiprocess env (pickle issues)
   - [x] bpe
   - [x] whisper
- [x]  unit test covers all member functions of tokenizer (可能还有没有看到的 之后有问题再看)
- [x] check related file
- [x] check training and recognize.py work

next pr

- [ ] paraformer tokenizer for fintune
- [ ] huggingface tokenizer for LLMs

### CODE REVIEW

**Comment by Mddct** [comment]
这里补充下whisper tokenizer：
原始官方实现中 tokenizer中对multi task的sot language等在post_init 中实现
image

在wenet中我们选择了，外边添加一个函数：
https://github.com/wenet-e2e/wenet/blob/main/wenet/utils/common.py#L150-L213

所以这里的whisper tokenizer需要fix下他填充的东西 
TODO
- [ ] fix whisper prepend

**Comment by xingchensong** [comment]
`post_init` 不用管他，不灵活，弃用。举个例子 ，多任务学习时，可能有 `transcribe` 任务，也可能有 `translate` 任务，如果执意用 `self.sot_sequence`, 那么要改成员属性

**Comment by Mddct** [comment]
> `post_init` 不用管他，不灵活，弃用。举个例子 ，多任务学习时，可能有 `transcribe` 任务，也可能有 `translate` 任务，如果执意用 `self.sot_sequence`, 那么要改成员属性

确认了下，pos_init 不影响其他的函数， 看着目前的实现中应该也没有问题了

File: wenet/dataset/processor.py [hunk 1]
```diff
@@ -26,9 +25,9 @@
 import torchaudio.compliance.kaldi as kaldi
 import torch.nn.functional as F
 from torch.nn.utils.rnn import pad_sequence
-from wenet.utils.tokenize_utils import tokenize_by_bpe_model
+from wenet.text.base_tokenizer import BaseTokenizer
 
-torchaudio.utils.sox_utils.set_buffer_size(16500)
+# torchaudio.utils.sox_utils.set_buffer_size(16500)
```
Comment (xingchensong) [inline wenet/dataset/processor.py:30]: 为啥注释掉了

Comment (Mddct) [reply wenet/dataset/processor.py:30]: 最后打开 本地mac有问题（mac上torchaudio 不支持sox backend）， 需要修复下

File: wenet/utils/common.py [hunk 2]
```diff
@@ -176,6 +173,8 @@ def add_whisper_tokens(
         ys_out (torch.Tensor) : (B, Lmax + ?)
 
     """
+    from whisper.tokenizer import LANGUAGES as WhiserLanguages
+    WHISPER_LANGS = tuple(WhiserLanguages.keys())
     if use_prev:
         # i.e., hotword list
         _prev = [tokenizer.sot_prev]
```
Comment (xingchensong) [inline wenet/utils/common.py:180]: 这个为啥挪到里面了

Comment (Mddct) [reply wenet/utils/common.py:180]: cli也可以不报错（现在cli直接装了个whisper）

Comment (Mddct) [reply wenet/utils/common.py:180]: 嗯 get ，之后这里有个TODO
- [ ] 取消条件引入

Comment (robin1001) [reply wenet/utils/common.py:180]: 条件引入会多次调用吗？

Comment (Mddct) [reply wenet/utils/common.py:180]: 应该会， 函数作用域里import 多次调用函数 会多次import，
包级import 只会有一次

**Comment by Mddct** [comment]
build sp model lazily , 

https://github.com/espnet/espnet/blob/master/espnet2/text/sentencepiece_tokenizer.py#L14-L17

**Comment by Mddct** [comment]
whisper tiktokenize 在多进程环境先会有问题， 所有也改成了lazy build的形式
issues：https://github.com/huggingface/datasets/issues/5769

**Comment by Mddct** [comment]
recognize.py works!
Screenshot 2023-11-28 at 19 41 57

**Comment by Mddct** [comment]
training works
Screenshot 2023-11-28 at 21 02 19

**Comment by Mddct** [comment]
@robin1001 @xingchensong   all work, it's time to merge

**Review by xingchensong — APPROVED** [review]

**Comment by robin1001** [comment]
Really a lot of work, and all look great! Just one question, seems `WenetTokenizer` is not initialized in `init_tokenzer.py`.

**Comment by Mddct** [comment]
> Really a lot of work, and all look great! Just one question, seems `WenetTokenizer` is not initialized in `init_tokenzer.py`.

Split WenetTokenizer into char and bpe， WenetTokenizer is just a tool to aligm the result for char and bpe tokenizer

**Comment by Mddct** [comment]
> > Really a lot of work, and all look great! Just one question, seems `WenetTokenizer` is not initialized in `init_tokenzer.py`.
> 
> Split WenetTokenizer into char and bpe， WenetTokenizer is just a tool to aligm the result for char and bpe tokenizer

Will delete in future pr

**Pull request merged**